lib/patch-by-id.js

```diff
@@ -0,0 +1,187 @@
+  }
+  assert(arguments.length >= 2, 'At least two arguments are required');
+
+  var Model = this;
+  var where = {};
+  var strict = this.settings.strict;
+  // TODO: the methods which we have imported from dao (idName(),
+  // stillConnecting(), isWhereByGivenId(),...), may need to be
+  // moved to a more appropriate file (utils.js?)
+  var idPropertyName = dao.idName(Model);
+
+  where[idPropertyName] = id;
+  if (cb === undefined) {
+    if (typeof options === 'function') {
+      // patchById(id, data, cb)
+      cb = options;
+      options = {};
+    }
+  }
+
+  data = data || {};
+  cb = cb || utils.createPromiseCallback();
+  options = options || {};
+
+  assert((typeof data === 'object') && (data !== null),
+    'The data argument must be an object');
+  assert(typeof options === 'object', 'The options argument must be an object');
+  assert(typeof cb === 'function', 'The cb argument must be a function');
+
+  var connector = Model.getDataSource().connector;
+  assert(typeof connector.update === 'function',
+    'update() must be implemented by the connector');
+
+  // Make sure id cannot be changed
+  checkId(data[idPropertyName], id, idPropertyName, cb);
+
+  // Question[1]:
+  // Please see QUESTION-test[1] in`test\dao.suit\patch-by-id.suits.js` for failing test
+  // data = data.toObject();
+  // data = removeUndefined(data);
+
+  var hookState = {};
+  this.applyProperties(data);
+
+  var context = {
+    Model: Model,
+    where: where,
+    data: data,
+    hookState: hookState,
+    options: options,
+  };
+
+  Model.notifyObserversOf('before save', context, function(err, ctx) {
+    if (err) return cb(err);
+    var isOriginalQuery = dao.isWhereByGivenId(Model, ctx.where, id);
+
+    if (!isOriginalQuery) {
+      var err = new Error('ctx.where:' + JSON.stringify(ctx.where) + ' is not valid; the `' +
+        idPropertyName + '` original value is ' + id);
+      err.statusCode = 403;
+      process.nextTick(function() {
+        cb(err);
+      });
+      return cb.promise;
+    }
+
+    if (strict) {
+      var instInfo = { '__unknownProperties': [] };
+      dao.applyStrictCheck(Model, strict, data,
+        instInfo, validateAndCallConnector);
+    } else {
+      validateAndCallConnector();
+    }
+
+    function validateAndCallConnector(err) {
+      if (err) return cb(err);
+      if (options.validate === false) {
+        return doUpdate(ctx.where, ctx.data);
+      }
+
+      // only when options.validate is not set, take model-setting into consideration
+      if (options.validate === undefined && Model.settings.automaticValidation === false) {
+        return doUpdate(ctx.where, ctx.data);
+      }
+
+      // Make sure id is not changed in the context of `before save`
+      checkId(ctx.data[idPropertyName], id, idPropertyName, cb);
+
+      dataObj = new Model(ctx.data);
+      dataObj.unsetAttribute(idPropertyName);
+      // validation required
+      dataObj.isValid(function(valid) {
```

Comment:
This will not work in most cases. Consider a model with required "name" and optional "age" properties, and then call `patchById` setting only `age:42`. The request is perfectly valid, but the validation will fail.

Ideally, we could run validators only on the properties affected by the patch operation. However, I think that's nontrivial and should be left out of scope of this pull request.

OTOH, because we are skipping validation, calling `patchById` makes it easy to get the data into inconsistent state. I think it may be better to not expose this method in the public REST API by default. In which case, we can probably left it out of 3.0 completely and wait until we have more time available to implement it fully with validations.

Thoughts?